### PR TITLE
Added support for activity update

### PR DESCRIPTION
**Pull request by sbawaskar**
Addresses #35.

### CODE REVIEW

**Comment by sbawaskar** [comment]
@PrasadG193 , please review.

**Comment by PrasadG193** [comment]
Thanks for the PR @sbawaskar. Will it be possible to add a sample for this under https://github.com/infracloudio/msbotbuilder-go/tree/develop/samples? You can add the same example you are using for testing

**Comment by sbawaskar** [comment]
> Thanks for the PR @sbawaskar. Will it be possible to add a sample for this under https://github.com/infracloudio/msbotbuilder-go/tree/develop/samples? You can add the same example you are using for testing

Will add a sample and test for that sample.

**Comment by sbawaskar** [comment]
@PrasadG193 , to update an activity we need ID of activity which was previously created by bot. Example, on an approval card, approval button is removed once user click on approval. With current implementation of SDK I did not find a way to get activity ID which bot created. Any insight on this?

**Comment by PrasadG193** [comment]
@sbawaskar We will have to store the activity ID which we want to update. E.g in proactive activity sample (https://github.com/infracloudio/msbotbuilder-go/blob/develop/samples/proactive/main.go#L89), we save conv ref when the user sends a message. Later we proactively send a welcome message to the same conversation using ref we store

**Comment by sbawaskar** [comment]
> @sbawaskar We will have to store the activity ID which we want to update. E.g in proactive activity sample (https://github.com/infracloudio/msbotbuilder-go/blob/develop/samples/proactive/main.go#L89), we save conv ref when the user sends a message. Later we proactively send a welcome message to the same conversation using ref we store

Conversation reference is saved and that works for proactive message. For activity update we need activity ID. Getting below error when I try to apply conversation reference.
```
https://smba.trafficmanager.net/in/v3/conversations/a:1lFpRAr-rUSU7muWrjctS2Cl1t0fHQLinwHybvsmNfzczNldZvwiDfDuQuCGpC7yftxsm8L-YCMTmz2xVyuxB2TZqpwI4HD7cHuPjoB6KfuqNix0AM2ZKznr0JekpD69Y/activities/1603215439486
{"error":{"code":"ServiceError","message":"Unknown"}}
```
Sample code,

```
activityInstance, err := ht.Adapter.ParseRequest(ctx, req)
	if err != nil {
		fmt.Println("Failed to parse request.", err)
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	err = ht.Adapter.ProcessActivity(ctx, activityInstance, customHandler)
	if err != nil {
		fmt.Println("Failed to process request", err)
		http.Error(w, err.Error(), http.StatusBadRequest)
		fmt.Println(err.Error())
		return
	}
	conversationRef := activity.GetCoversationReference(activityInstance)
	act := activity.ApplyConversationReference(schema.Activity{Type: schema.Message}, conversationRef, true)
	if activityID != "" {
		act.Text = "Changed Activity"
		act.ID = activityID
		err = ht.Adapter.UpdateActivity(ctx, act)
		if err != nil {
			fmt.Println("Failed to process request", err)
			http.Error(w, err.Error(), http.StatusBadRequest)
			fmt.Println(err.Error())
			return
		}
                // activityID is a global variable. We save previous posted activity ID and keep on updating it.
		activityID = conversationRef.ActivityID
	} else {
		activityID = conversationRef.ActivityID
	}
	fmt.Println("Request processed successfully.")
```

**Comment by sbawaskar** [comment]
@PrasadG193 , Added sample for activity update. Please check.

**Comment by sbawaskar** [comment]
@PrasadG193 , reminder for review.

**Comment by sbawaskar** [comment]
@PrasadG193 , reminder for review.

**Comment by PrasadG193** [comment]
Thank you @sbawaskar. Please give me some time, I will try to review it tomorrow.

File: connector/client/client.go [hunk 1]
```diff
@@ -87,6 +89,24 @@ func (client *ConnectorClient) Delete(target url.URL, activity schema.Activity)
...
+// Put an activity.
+//
+// Creates a HTTP PUT request with the provided activity payload and a Bearer token in the header.
+// Returns any error as received from the call to connector service.
+func (client *ConnectorClient) Put(target url.URL, activity schema.Activity) error {
+	jsonStr, err := json.Marshal(activity)
+	if err != nil {
+		return err
+	}
+	fmt.Println(target.String())
```
Comment (PrasadG193) [inline connector/client/client.go:101]: Debug statement? Can we remove it or log it in debug mode?

File: connector/client/client.go [hunk 2]
```diff
@@ -109,6 +129,8 @@ func (client *ConnectorClient) checkRespError(resp *http.Response, err error) er
 		}
 	}
 	defer resp.Body.Close()
+	sBody, err := ioutil.ReadAll(resp.Body)
```
Comment (PrasadG193) [inline connector/client/client.go:132]: Please handle returned error

File: connector/client/client.go [hunk 3]
```diff
@@ -109,6 +129,8 @@ func (client *ConnectorClient) checkRespError(resp *http.Response, err error) er
...
+	fmt.Print(string(sBody))
```
Comment (PrasadG193) [inline connector/client/client.go:133]: Do we need to print this?

File: core/activity/response.go [hunk 4]
```diff
@@ -33,15 +33,15 @@ import (
...
 }
 
 const (
 	// APIVersion for response URLs
 	APIVersion = "v3"
 
 	sendToConversationURL = "/%s/conversations/%s/activities"
-	replyToActivityURL    = "/%s/conversations/%s/activities/%s"
-	deleteActivityURL     = "/%s/conversations/%s/activities/%s"
+	activityResourceURL   = "/%s/conversations/%s/activities/%s"
```
Comment (PrasadG193) [inline core/activity/response.go:44]: :+1:

File: connector/client/client.go (hunk 1 above)
Comment (sbawaskar) [reply connector/client/client.go:101]: Sure will move it to debug.

File: connector/client/client.go (hunk 3 above)
Comment (sbawaskar) [reply connector/client/client.go:133]: Sure will move it to debug.

**Comment by sbawaskar** [comment]
@PrasadG193 , addressed code review comments, please review.

File: samples/activity-update/README.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+You can either test BOT on BotFramework portal or you can create app manifest and install the App on Teams as mentioned [here](https://docs.microsoft.com/en-us/microsoftteams/platform/bots/how-to/create-a-bot-for-teams#create-your-app-manifest-and-package).
+
+
+## Understanding the example
+
+The program starts by creating a handler struct of type `activity.HandlerFuncs`.
+
+This struct contains definition for the `OnMessageFunc` field which is a treated as a callback by the library on the respective event.
+In this sample when GetCard command is issued from chat application adaptive card with input box is sent from this bot. Once submit button on card is clicked the card is deleted.
```
Comment (PrasadG193) [inline samples/activity-update/README.md:48]: I believe the application also update the text on the card - "Changed Activity" which demonstrates the Update activity handling. Please document this behavior.

Comment (PrasadG193) [reply samples/activity-update/README.md:48]: @sbawaskar please let me know once ready

Comment (sbawaskar) [reply samples/activity-update/README.md:48]: Sure

Comment (sbawaskar) [reply samples/activity-update/README.md]: Done thanks!

Comment (sbawaskar) [reply samples/activity-update/README.md]: @PrasadG193 , updated doc. Please check.

File: samples/activity-update/README.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+You can either test BOT on BotFramework portal or you can create app manifest and install the App on Teams as mentioned [here](https://docs.microsoft.com/en-us/microsoftteams/platform/bots/how-to/create-a-bot-for-teams#create-your-app-manifest-and-package).
+
+
+## Understanding the example
+
+The program starts by creating a handler struct of type `activity.HandlerFuncs`.
+
+This struct contains definition for the `OnMessageFunc` field which is a treated as a callback by the library on the respective event.
+In this sample when GetCard command is issued from chat application adaptive card with input box is sent from this bot. Once submit button on card is clicked the card is deleted. Also bot updates the activity and set text message to activity as "Changed Activity" after card is deleted.
```
Comment (PrasadG193) [inline samples/activity-update/README.md:48]:
```suggestion
This struct contains a definition for the `OnMessageFunc` field which is treated as a callback by the library on the respective event.
In this sample, when the GetCard command is issued from the chat application, an adaptive card with an input box is sent from this bot. Once submit button on the card is clicked the card is deleted. Also, the bot updates the activity and sets the text message to activity as "Changed Activity" after the card is deleted.
```

**Review by PrasadG193 — APPROVED** [review]

**Pull request merged**